File: app/src/main/java/com/ndmsystems/coala/message/CoAPMessage.java
package com.ndmsystems.coala.message;

import com.ndmsystems.coala.helpers.Hex;
import com.ndmsystems.coala.helpers.MessageHelper;
import com.ndmsystems.coala.helpers.StringHelper;
import com.ndmsystems.coala.layers.response.ResponseHandler;
import com.ndmsystems.infrastructure.logging.LogHelper;

import java.io.UnsupportedEncodingException;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;

public class CoAPMessage {

    public static void convertToEmptyAck(CoAPMessage message,
                                         InetSocketAddress from) {
        message.setType(CoAPMessageType.ACK);
        message.setCode(CoAPMessageCode.CoapCodeEmpty);
        message.setAddress(from);
        message.setPayload(null);
    }

    public static CoAPMessage ackTo(CoAPMessage message,
                                    InetSocketAddress from,
                                    CoAPMessageCode code) {
        CoAPMessage result = new CoAPMessage(CoAPMessageType.ACK, code);
        result.setId(message.getId());
        result.setToken(message.getToken());
        result.setURIScheme(message.getURIScheme());
        result.setAddress(from);
        CoAPMessageOption option = message.getOption(CoAPMessageOptionCode.OptionObserve);
        if (option != null)
            result.addOption(option);
        return result;
    }

    public static CoAPMessage resetTo(CoAPMessage message, InetSocketAddress from) {
        CoAPMessage result = new CoAPMessage(CoAPMessageType.RST, CoAPMessageCode.CoapCodeEmpty);
        result.setId(message.getId());
        result.setToken(message.getToken());
        result.setURIScheme(message.getURIScheme());
        result.setAddress(from);
        return result;
    }


    private InetSocketAddress destination;

    private int id;
    private CoAPMessageType type;
    private CoAPMessageCode code;
    private CoAPMessagePayload payload;
    private List<CoAPMessageOption> options = new ArrayList<>();
    private byte[] token = null;
    private InetSocketAddress proxy;
    private ResponseHandler responseHandler;
    private ResendHandler resendHandler = () -> LogHelper.v("Resend message: " + id);
    private byte[] peerPublicKey = null;

    public CoAPMessage(CoAPMessage message) {
        this(message.getType(), message.getCode(), message.getId());
        if (message.getPayload() != null) {
            payload = new CoAPMessagePayload(message.toString());
        }
        options = new ArrayList<>(message.getOptions());
        if (message.getToken() != null) {
            token = new byte[message.getToken().length];
            System.arraycopy(message.getToken(), 0, token, 0, token.length);
        }

        if (message.getProxy() != null) this.proxy = message.getProxy();
        if (message.destination != null) this.destination = message.destination;
        if (message.responseHandler != null) this.responseHandler = message.responseHandler;
        if (message.resendHandler != null) this.resendHandler = message.resendHandler;
        if (message.peerPublicKey != null) this.peerPublicKey = message.peerPublicKey;
    }

    public CoAPMessage(CoAPMessageType messageType, CoAPMessageCode messageCode) {
        this(messageType, messageCode, MessageHelper.generateId());
    }

    public CoAPMessage(CoAPMessageType messageType, CoAPMessageCode messageCode, int id) {
        this.type = messageType;
        this.code = messageCode;
        this.id = id;
    }

    public ResponseHandler getResponseHandler() {
        return responseHandler;
    }

    public void setResponseHandler(ResponseHandler responseHandler) {
        this.responseHandler = responseHandler;
    }

    public boolean isRequest() {
        return code.isRequest();
    }

    public int getId() {
        return id;
    }

    public CoAPMessageType getType() {
        return type;
    }

    public CoAPMessageCode getCode() {
        return code;
    }

    public byte[] getToken() {
        return token;
    }

    public String getHexToken() {
        return Hex.encodeHexString(token);
    }


    public CoAPMessage setId(int id) {
        this.id = id;
        return this;
    }

    public CoAPMessage setType(CoAPMessageType type) {
        this.type = type;
        return this;
    }

    public CoAPMessage setCode(CoAPMessageCode code) {
        this.code = code;
        return this;
    }

    public CoAPMessage setToken(byte[] token) {
        this.token = token;
        return this;
    }

    public CoAPRequestMethod getMethod() {
        switch (this.code) {
            case GET:
                return CoAPRequestMethod.GET;
            case POST:
                return CoAPRequestMethod.POST;
            case PUT:
                return CoAPRequestMethod.PUT;
            case DELETE:
                return CoAPRequestMethod.DELETE;

            default:
                return null;

        }
    }

    public void setOptions(List<CoAPMessageOption> options) {
        this.options = new ArrayList<>(options);
    }

    public List<CoAPMessageOption> getOptions() {
        return new ArrayList<>(options);
    }

    public boolean hasOption(CoAPMessageOptionCode code) {
        return getOption(code) != null;
    }

    public CoAPMessagePayload getPayload() {
        return payload;
    }

    public CoAPMessage setPayload(CoAPMessagePayload payload) {
        this.payload = payload;
        return this;
    }

    public CoAPMessage setStringPayload(String payload) {
        if (this.payload == null) {
            this.payload = new CoAPMessagePayload(payload.getBytes());
        } else {
            this.payload.content = payload.getBytes();
        }
        return this;
    }

    public void setDestination(InetSocketAddress address) {
        this.destination = address;
    }

    public InetSocketAddress getDestination() {
        if (destination != null) return destination;
        return getAddress();
    }

    /**
     * Example: <code>[scheme]://[host]:[port]{/resource}*?{&amp;query}*</code>
     *
     * @return CoAPMessage
     * @throws IllegalArgumentException
     */
    public CoAPMessage setURI(String uri) {
        try {
            if (!uri.startsWith(Scheme.NORMAL.toString() + "://") && !uri.startsWith(Scheme.SECURE.toString() + "://")) {
                uri = Scheme.NORMAL.toString() + "://" + uri;
            }
            return setURI(new URI(uri));
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException("Failed to set uri " + uri + ": " + e.getMessage());
        }
    }

    public CoAPMessage setURI(URI uri) {
        setURIScheme(Scheme.fromString(uri.getScheme()));

        setURIHost(uri.getHost());

        setURIPort(uri.getPort());

        setURIPath(uri.getPath());

        setURIQuery(uri.getQuery());

        return this;
    }

    public String getURI() {
        StringBuilder builder = new StringBuilder();

        if (hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
            builder.append((String) getOption(CoAPMessageOptionCode.OptionProxyURI).value);
        } else {
            builder.append(getURIScheme().toString()).append("://")
                    .append(getURIHost()).append(":").append(getURIPort());
        }

        String uriPath = getURIPathString();
        if (uriPath != null) {
            builder.append("/" + uriPath);
        }

        List<CoAPMessageOption> queryOptions = this.getOptions(CoAPMessageOptionCode.OptionURIQuery);
        if (queryOptions.size() > 0) {
            builder.append("?");
            for (CoAPMessageOption parameter : queryOptions) {
                String parameterString = (String) parameter.value;
                int index = parameterString.indexOf("=");
                String key = parameterString.substring(0, index);
                String value = parameterString.substring(index + 1);
                try {
                    value = URLEncoder.encode(value, "UTF-8").replaceAll("[+]", "%20");
                } catch (UnsupportedEncodingException ignore) {
                    LogHelper.e("Can't encode query parameter: " + value);
                }
                builder.append(key + "=" + value + "&");
            }
            builder.setLength(builder.length() - 1);
        }

        return builder.toString();
    }

    public String getURIHost() {
        CoAPMessageOption option = this.getOption(CoAPMessageOptionCode.OptionURIHost);

        if (option == null) {
            if (destination != null) {
                return destination.getAddress().getHostAddress();
            }
            return "localhost";
        }

        return (String) option.value;
    }

    public CoAPMessage setURIHost(String host) {
        CoAPMessageOption option = this.getOption(CoAPMessageOptionCode.OptionURIHost);

        if (option == null) {
            option = new CoAPMessageOption(CoAPMessageOptionCode.OptionURIHost, host);
        } else option.value = host;

        addOption(option);
        return this;
    }

    public Integer getURIPort() {
        CoAPMessageOption option = this.getOption(CoAPMessageOptionCode.OptionURIPort);

        if (option == null) {
            if (destination != null) {
                return destination.getPort();
            }
            return 5683;
        }

        return (Integer) option.value;
    }

    public CoAPMessage setURIPort(int port) {
        CoAPMessageOption option = this.getOption(CoAPMessageOptionCode.OptionURIPort);

        if (option == null) {
            option = new CoAPMessageOption(CoAPMessageOptionCode.OptionURIPort, port);
        }
        option.value = port;

        addOption(option);
        return this;
    }

    public String getURIPathString() {
        List<CoAPMessageOption> pathOptions = this.getOptions(CoAPMessageOptionCode.OptionURIPath);

        if (pathOptions.size() == 0) {
            return null;
        }

        List<String> pathParts = new ArrayList<String>();
        for (CoAPMessageOption pathElem : pathOptions) {
            pathParts.add((String) pathElem.value);
        }

        return StringHelper.join(pathParts, "/");
    }

    public String getURIQueryString() {
        List<CoAPMessageOption> queryOptions = this.getOptions(CoAPMessageOptionCode.OptionURIQuery);

        if (queryOptions.size() == 0) {
            return null;
        }

        List<String> queryParts = new ArrayList<String>();

        for (CoAPMessageOption queryElem : queryOptions) {
            queryParts.add((String) queryElem.value);
        }

        return StringHelper.join(queryParts, "&");
    }

    public String getURIQuery(String key) {
        List<CoAPMessageOption> queryOptions = this.getOptions(CoAPMessageOptionCode.OptionURIQuery);

        for (CoAPMessageOption queryElem : queryOptions) {
            String[] parts = ((String) queryElem.value).split("=");
            if (parts.length == 2 && parts[0].equals(key)) {
                return parts[1];
            }
        }

        return "";
    }

    public CoAPMessage addQueryParam(String key, String value) {
        addOption(new CoAPMessageOption(CoAPMessageOptionCode.OptionURIQuery, key + "=" + value));
        return this;
    }

    public CoAPMessage addQueryParams(Map<String, String> params) {
        if (params != null)
            for (String key : params.keySet())
                addQueryParam(key, params.get(key));
        return this;
    }

    public void setURIQuery(String query) {
        if (query != null) {
            StringTokenizer st = new StringTokenizer(query, "&");
            while (st.hasMoreTokens()) {
                String[] parts = st.nextToken().split("=");
                if (parts.length < 2)
                    LogHelper.e("Wrong parts");
                else
                    addQueryParam(parts[0], parts[1]);
            }
        }
    }

    public CoAPMessage setURIScheme(Scheme scheme) {
        CoAPMessageOption option = this.getOption(CoAPMessageOptionCode.OptionURIScheme);

        if (option == null) {
            option = new CoAPMessageOption(CoAPMessageOptionCode.OptionURIScheme, scheme.toInt());
        } else option.value = scheme.toInt();

        addOption(option);
        return this;
    }

    public Scheme getURIScheme() {
        CoAPMessageOption option = this.getOption(CoAPMessageOptionCode.OptionURIScheme);

        if (option == null) {
            return Scheme.NORMAL;
        }

        return Scheme.fromInt((Integer) option.value);
    }

    public CoAPMessage setURIPath(String path) {
        removeOption(CoAPMessageOptionCode.OptionURIPath);
        if (path != null) {
            String[] pathSegments = path.split("/");
            for (String pathSegment : pathSegments)
                if (!pathSegment.isEmpty())
                    addOption(new CoAPMessageOption(CoAPMessageOptionCode.OptionURIPath, pathSegment));
        }
        return this;
    }

    /**
     * Returns the whole list of Options.
     * Used for repeatable options (such as URI-Query)
     *
     * @param optionCode
     * @return List<CoAPMessageOption>
     */
    public List<CoAPMessageOption> getOptions(CoAPMessageOptionCode optionCode) {
        List<CoAPMessageOption> options = new ArrayList<CoAPMessageOption>();

        for (CoAPMessageOption option : this.options) {
            if (option.code == optionCode) {
                options.add(option);
            }
        }

        return options;
    }

    /**
     * Returns the first occurrence of the requested option.
     *
     * @param optionCode
     * @return
     */
    public CoAPMessageOption getOption(CoAPMessageOptionCode optionCode) {
        for (CoAPMessageOption option : this.options) {
            if (option.code == optionCode) {
                return option;
            }
        }

        return null;
    }

    public InetSocketAddress getProxy() {
        return proxy;
    }

    public void setProxy(InetSocketAddress address) {
        if (address == null) return;
        proxy = address;
        CoAPMessageOption option = new CoAPMessageOption(CoAPMessageOptionCode.OptionProxyURI, getURIScheme().toString() +"://" + getURIHost() + ":" + getURIPort().toString());
        addOption(option);
    }

    public void removeOption(CoAPMessageOptionCode optionCode) {
        Iterator<CoAPMessageOption> iterator = options.listIterator();
        while (iterator.hasNext()) {
            CoAPMessageOption option = iterator.next();
            if (option.code == optionCode)
                iterator.remove();
        }
    }

    public CoAPMessage addOption(CoAPMessageOption option) {
        if (option == null)
            return this;

        if (!option.isRepeatable())
            removeOption(option.code);

        options.add(option);
        return this;
    }

    public String toString() {
        if (this.payload == null) {
            return "";
        }
        return this.payload.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CoAPMessage message = (CoAPMessage) o;
        return id == message.id;
    }

    @Override
    public int hashCode() {
        return id;
    }

    public void setAddress(InetSocketAddress address) {
        if (address.getAddress() != null
                && address.getAddress().getHostAddress() != null) {
            setURIHost(address.getAddress().getHostAddress());
        } else {
            setURIHost(address.getHostName());
        }
        setURIPort(address.getPort());
    }

    public InetSocketAddress getAddress() {
        return new InetSocketAddress(getURIHost(), getURIPort());
    }

    public void setResendHandler(ResendHandler resendHandler) {
        this.resendHandler = resendHandler;
    }

    public ResendHandler getResendHandler() {
        return resendHandler;
    }

    public byte[] getPeerPublicKey() {
        return peerPublicKey;
    }

    public void setPeerPublicKey(byte[] peerPublicKey) {
        this.peerPublicKey = peerPublicKey;
    }

    public enum MediaType {
        TextPlain(0),
        LinkFormat(40),
        Xml(41),
        OctetStream(42),
        Exi(47),
        Json(50);

        private final int value;

        MediaType(int value) {
            this.value = value;
        }

        public static MediaType fromInt(Integer value) {
            switch (value) {
                case 0:
                    return TextPlain;
                case 40:
                    return LinkFormat;
                case 41:
                    return Xml;
                case 42:
                    return OctetStream;
                case 47:
                    return Exi;
                case 50:
                    return Json;
                default:
                    LogHelper.e("Unknown MediaType");
                    return null;
            }
        }

        public int toInt() {
            return value;
        }
    }

    public enum Scheme {
        NORMAL, SECURE;

        public Integer toInt() {
            switch (this) {
                case SECURE:
                    return 1;
                default:
                    return 0;
            }
        }

        public String toString() {
            switch (this) {
                case SECURE:
                    return "coaps";
                default:
                    return "coap";
            }
        }

        public static Scheme fromInt(Integer value) {
            switch (value) {
                case 1:
                    return SECURE;
                default:
                    return NORMAL;
            }
        }

        public static Scheme fromString(String value) {
            switch (value) {
                case "coaps":
                case "1":
                    return SECURE;
                default:
                    return NORMAL;
            }
        }
    }

    public static interface ResendHandler {
        public void onResend();
    }
}

File: app/src/main/java/com/ndmsystems/coala/message/CoAPMessage.kt
package com.ndmsystems.coala.message

import com.ndmsystems.coala.Coala
import com.ndmsystems.coala.helpers.Hex.encodeHexString
import com.ndmsystems.coala.helpers.MessageHelper.generateId
import com.ndmsystems.coala.helpers.StringHelper.join
import com.ndmsystems.coala.helpers.logging.LogHelper.e
import com.ndmsystems.coala.helpers.logging.LogHelper.v
import com.ndmsystems.coala.helpers.logging.LogHelper.w
import com.ndmsystems.coala.layers.response.ResponseHandler
import java.io.UnsupportedEncodingException
import java.net.InetSocketAddress
import java.net.URI
import java.net.URISyntaxException
import java.net.URLEncoder
import java.util.StringTokenizer

class CoAPMessage @JvmOverloads constructor(var type: CoAPMessageType, var code: CoAPMessageCode, var id: Int = generateId()) {
    lateinit var address: InetSocketAddress
    var proxy: InetSocketAddress? = null
        private set
    var payload: CoAPMessagePayload? = null
    private var options: MutableList<CoAPMessageOption> = ArrayList()
    var token: ByteArray? = null
    var responseHandler: ResponseHandler? = null
    var resendHandler: ResendHandler = object: ResendHandler {
        override fun onResend() {
            v("Resend message: $id")
        }
    }
    var peerPublicKey: ByteArray? = null
    var isRequestWithLongTimeNoAnswer = false

    constructor(message: CoAPMessage) : this(message.type, message.code, message.id) {
        if (message.payload != null) {
            payload = CoAPMessagePayload(message.payload!!.content)
        }
        options = ArrayList(message.getOptions())
        if (message.token != null) {
            token = ByteArray(message.token!!.size)
            System.arraycopy(message.token!!, 0, token, 0, token!!.size)
        }
        if (message.proxy != null) proxy = message.proxy
        address = message.address
        if (message.responseHandler != null) responseHandler = message.responseHandler
        resendHandler = message.resendHandler
        if (message.peerPublicKey != null) peerPublicKey = message.peerPublicKey
        isRequestWithLongTimeNoAnswer = message.isRequestWithLongTimeNoAnswer
    }

    val isRequest: Boolean
        get() = code.isRequest
    val hexToken: String
        get() = encodeHexString(token)

    fun setId(id: Int): CoAPMessage {
        this.id = id
        return this
    }

    fun setType(type: CoAPMessageType): CoAPMessage {
        this.type = type
        return this
    }

    fun setCode(code: CoAPMessageCode): CoAPMessage {
        this.code = code
        return this
    }

    val method: CoAPRequestMethod?
        get() = when (this.code) {
            CoAPMessageCode.GET -> CoAPRequestMethod.GET
            CoAPMessageCode.POST -> CoAPRequestMethod.POST
            CoAPMessageCode.PUT -> CoAPRequestMethod.PUT
            CoAPMessageCode.DELETE -> CoAPRequestMethod.DELETE
            else -> null
        }

    fun setOptions(options: List<CoAPMessageOption>) {
        this.options = ArrayList(options)
    }

    fun getOptions(): List<CoAPMessageOption> {
        return ArrayList(options)
    }

    fun hasOption(code: CoAPMessageOptionCode): Boolean {
        return getOption(code) != null
    }

    fun setStringPayload(payload: String): CoAPMessage {
        if (this.payload == null) {
            this.payload = CoAPMessagePayload(payload.toByteArray())
        } else {
            this.payload!!.content = payload.toByteArray()
        }
        return this
    }

    /**
     * Example: `[scheme]://[host]:[port]{/resource}*?{&query}*`
     *
     * @return CoAPMessage
     */
    fun setURI(uri: String): CoAPMessage {
        var uri = uri
        return try {
            if (!uri.startsWith(Scheme.NORMAL.toString() + "://") && !uri.startsWith(Scheme.SECURE.toString() + "://")) {
                uri = Scheme.NORMAL.toString() + "://" + uri
            }
            setURI(URI(uri))
        } catch (e: URISyntaxException) {
            throw IllegalArgumentException("Failed to set uri " + uri + ": " + e.message)
        }
    }

    fun setURI(uri: URI): CoAPMessage {
        setURIScheme(Scheme.fromString(uri.scheme))
        removeOption(CoAPMessageOptionCode.OptionURIHost)
        removeOption(CoAPMessageOptionCode.OptionURIPort)
        val port = if (uri.port != -1) uri.port else Coala.DEFAULT_PORT
        address = InetSocketAddress(uri.host, port)
        uri.path?.let { setURIPath(it) }
        uri.query?.let { setURIQuery(it) }
        return this
    }

    fun getURI(): String {
        val builder = StringBuilder()
        val port: Int = if (address.port != -1) address.port else Coala.DEFAULT_PORT
        val host: String = if (address.address != null && address.address.hostAddress != null) address.address.hostAddress else {
            w("Address is null! return \"null\"")
            "null"
        }
        if (hasOption(CoAPMessageOptionCode.OptionProxyURI)) {
            builder.append(getOption(CoAPMessageOptionCode.OptionProxyURI)!!.value as String?)
        } else {
            builder.append(getURIScheme().toString()).append("://").append(host).append(":").append(port)
        }
        val uriPath = getURIPathString()
        builder.append("/").append(uriPath)

        val queryOptions = this.getOptions(CoAPMessageOptionCode.OptionURIQuery)
        if (queryOptions.isNotEmpty()) {
            builder.append("?")
            for (parameter in queryOptions) {
                val parameterString = parameter.value as String?
                val index = parameterString!!.indexOf("=")
                val key = parameterString.substring(0, index)
                var value = parameterString.substring(index + 1)
                try {
                    value = URLEncoder.encode(value, "UTF-8").replace("[+]".toRegex(), "%20")
                } catch (ignore: UnsupportedEncodingException) {
                    e("Can't encode query parameter: $value")
                }
                builder.append(key).append("=").append(value).append("&")
            }
            builder.setLength(builder.length - 1)
        }
        return builder.toString()
    }

    fun getURIPathString(): String {
        val pathOptions = this.getOptions(CoAPMessageOptionCode.OptionURIPath)
        if (pathOptions.isEmpty()) {
            return ""
        }
        val pathParts: MutableList<String?> = java.util.ArrayList()
        for (pathElem in pathOptions) {
            pathParts.add(pathElem.value as String?)
        }
        return join(pathParts, "/")
    }

    fun getURIQuery(key: String): String {
        val queryOptions = this.getOptions(CoAPMessageOptionCode.OptionURIQuery)
        for (queryElem in queryOptions) {
            val parts = (queryElem.value as String).split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (parts.size == 2 && parts[0] == key) {
                return parts[1]
            }
        }
        return ""
    }

    fun addQueryParam(key: String, value: String?): CoAPMessage {
        addOption(CoAPMessageOption(CoAPMessageOptionCode.OptionURIQuery, "$key=$value"))
        return this
    }

    fun addQueryParams(params: Map<String, String>): CoAPMessage {
        for (key in params.keys) addQueryParam(key, params[key])
        return this
    }

    private fun setURIQuery(query: String) {
        val st = StringTokenizer(query, "&")
        while (st.hasMoreTokens()) {
            val parts = st.nextToken().split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (parts.size < 2) e("Wrong parts") else addQueryParam(parts[0], parts[1])
        }
    }

    fun setURIScheme(scheme: Scheme): CoAPMessage {
        var option = getOption(CoAPMessageOptionCode.OptionURIScheme)
        if (option == null) {
            option = CoAPMessageOption(CoAPMessageOptionCode.OptionURIScheme, scheme.toInt())
        } else option.value = scheme.toInt()
        addOption(option)
        return this
    }

    fun getURIScheme(): Scheme {
        val option = getOption(CoAPMessageOptionCode.OptionURIScheme) ?: return Scheme.NORMAL
        return Scheme.fromInt(option.value as Int?)
    }

    fun setURIPath(path: String): CoAPMessage {
        removeOption(CoAPMessageOptionCode.OptionURIPath)
        val pathSegments = path.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (pathSegment in pathSegments) if (pathSegment.isNotEmpty()) addOption(
            CoAPMessageOption(
                CoAPMessageOptionCode.OptionURIPath,
                pathSegment
            )
        )
        return this
    }

    /**
     * Returns the whole list of Options.
     * Used for repeatable options (such as URI-Query)
     *
     * @param optionCode
     * @return List<CoAPMessageOption>
    </CoAPMessageOption> */
    private fun getOptions(optionCode: CoAPMessageOptionCode): List<CoAPMessageOption> {
        val options: MutableList<CoAPMessageOption> = ArrayList()
        for (option in this.options) {
            if (option.code == optionCode) {
                options.add(option)
            }
        }
        return options
    }

    /**
     * Returns the first occurrence of the requested option.
     *
     * @param optionCode
     * @return
     */
    fun getOption(optionCode: CoAPMessageOptionCode): CoAPMessageOption? {
        for (option in options) {
            if (option.code == optionCode) {
                return option
            }
        }
        return null
    }

    fun setProxy(proxyAddress: InetSocketAddress) {
        proxy = proxyAddress
        val option = CoAPMessageOption(
            CoAPMessageOptionCode.OptionProxyURI,
            getURIScheme().toString() + "://" + address.address.hostAddress + ":" + address.port
        )
        addOption(option)
    }

    fun removeOption(optionCode: CoAPMessageOptionCode) {
        val iterator: MutableIterator<CoAPMessageOption> = options.listIterator()
        while (iterator.hasNext()) {
            val option = iterator.next()
            if (option.code == optionCode) iterator.remove()
        }
    }

    fun addOption(option: CoAPMessageOption): CoAPMessage {
        if (!option.isRepeatable) removeOption(option.code)
        options.add(option)
        return this
    }

    override fun toString(): String {
        return if (payload == null) {
            ""
        } else payload.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val message = other as CoAPMessage
        return id == message.id
    }

    override fun hashCode(): Int {
        return id
    }

    fun getProxySecurityId(): Long? {
        val option = getOption(CoAPMessageOptionCode.OptionProxySecurityID) ?: return null
        return option.value as Long?
    }

    fun setProxySecurityId(proxyId: Long): CoAPMessage {
        var option = getOption(CoAPMessageOptionCode.OptionProxySecurityID)
        if (option == null) {
            option = CoAPMessageOption(CoAPMessageOptionCode.OptionProxySecurityID, proxyId)
        }
        option.value = proxyId
        addOption(option)
        return this
    }

    enum class MediaType(private val value: Int) {
        TextPlain(0), LinkFormat(40), Xml(41), OctetStream(42), Exi(47), Json(50);

        fun toInt(): Int {
            return value
        }

        companion object {
        }
    }

    enum class Scheme {
        NORMAL, SECURE;

        fun toInt(): Int {
            return if (this == SECURE) {
                1
            } else 0
        }

        override fun toString(): String {
            return if (this == SECURE) {
                "coaps"
            } else "coap"
        }

        companion object {
            fun fromInt(value: Int?): Scheme {
                return if (value == 1) {
                    SECURE
                } else NORMAL
            }

            fun fromString(value: String?): Scheme {
                return when (value) {
                    "coaps", "1" -> SECURE
                    else -> NORMAL
                }
            }
        }
    }

    interface ResendHandler {
        fun onResend()
    }

    companion object {
        fun convertToEmptyAck(
            message: CoAPMessage,
            from: InetSocketAddress
        ) {
            message.setType(CoAPMessageType.ACK)
            message.setCode(CoAPMessageCode.CoapCodeEmpty)
            message.address = from
            message.payload = null
        }

        fun ackTo(
            message: CoAPMessage,
            from: InetSocketAddress,
            code: CoAPMessageCode
        ): CoAPMessage {
            val result = CoAPMessage(CoAPMessageType.ACK, code)
            result.setId(message.id)
            result.token = message.token
            result.setURIScheme(message.getURIScheme())
            result.address = from
            val option = message.getOption(CoAPMessageOptionCode.OptionObserve)
            if (option != null) result.addOption(option)
            val proxySecurityId = message.getOption(CoAPMessageOptionCode.OptionProxySecurityID)
            if (proxySecurityId != null) result.addOption(proxySecurityId)
            return result
        }

        fun resetTo(message: CoAPMessage, from: InetSocketAddress): CoAPMessage {
            val result = CoAPMessage(CoAPMessageType.RST, CoAPMessageCode.CoapCodeEmpty)
            result.setId(message.id)
            result.token = message.token
            result.setURIScheme(message.getURIScheme())
            result.address = from

            return result
        }
    }
}